### PR TITLE
feat: introduce ptv3 multihead model, rework framework

### PR DESCRIPTION
**Pull request by amadeuszsz**
## Summary

Adds 3D object detection on top of the PointTransformerV3 (PTv3) backbone with an interchangeable detection-head design. The same PTv3 backbone can drive either a CenterPoint-style dense head (`CenterHead`) or a TransFusion-style query head (`TransFusionHead`), selected purely from config. To support this cleanly the PR also reworks large parts of the framework: a shared PTv3 base model, a detection model base with epoch-level mAP, a detection datamodule with repeat-factor sampling, 3D box transforms, a per-module ONNX/TensorRT export path, and numerous framework-wide cleanups.

---

## Headline: PTv3 multi-head 3D detection

- **`PTv3DetectionModel`** (`models/detection3d/ptv3.py`): PTv3 backbone, then trainable BEV projection (`PTv3BEVProjection`, scatter of decoded point features onto a dense BEV canvas), then a BEV refinement encoder (`PTv3BEVEncoder` with residual blocks), then the detection head. Supports `freeze_backbone` (eval-locked) for head-only fine-tuning from a segmentation checkpoint.
- **Detection heads** (`models/detection3d/heads/`):
- `CenterHead`: CenterPoint-style dense heatmap + regression (reg/height/dim/rot/vel), Gaussian-focal heatmap loss, circle-NMS decode.
- `TransFusionHead`: query decoder (`TransFusionDecoderLayer`, `LearnedPositionalEncoding`), Hungarian one-to-one matching, auxiliary dense heatmap, and `ExportableMultiheadAttention` (export-safe attention that avoids the fused-MHA decomposition TensorRT mishandles).
- **Task modules** (`models/detection3d/task_modules/`): `HungarianAssigner3D`, `TransFusionBBoxCoder`, heatmap utilities (`gaussian_radius`, `draw_heatmap_gaussian`, `circle_nms`), and match costs (classification, BBox-BEV-L1, IoU3D).
- **Losses** (`losses/detection3d/`): `SigmoidFocalLoss`, `GaussianFocalLoss`.
- **Metric** (`metrics/detection3d.py`): `CenterDistanceMeanAP`, a nuScenes-style center-distance mAP with per-class distance caps (`eval_class_range`), attribute-based GT exclusions (`filter_attributes`), min-point filtering, per-range buckets, and a DDP-gathered total metric.
- **Configs** (`configs/tasks/detection3d/ptv3/`): `base`, `tiny`, and four leaf configs (`centerhead`/`transhead` x `voxel005_128m`/`voxel012_122m`, j6gen2 and nuscenes).

---

### Model base hierarchy
- **`BaseModel`** (`models/base.py`): `training_step`/`validation_step`/`test_step`/`predict_step` are marked `@final` to enforce the lifecycle contract. `_shared_step` now returns `(metrics, raw_outputs)`, and the eval steps return `{**metrics, "model_outputs": ...}` so post-step hooks can consume raw outputs without overriding the step.
- **`PTv3BaseModel`** (`models/segmentation3d/ptv3_base.py`): shared owner of the PTv3 backbone, grid/range geometry, `freeze_backbone`, export geometry, and the serialized backbone-export helper plus export modules. Both the PTv3 segmentation and detection models now inherit it (the segmentation model was refactored accordingly).
- **`Detection3DBaseModel`** (`models/detection3d/base.py`): adds epoch-level AP accumulation via Lightning `on_*_batch_end`/`on_*_epoch_end` hooks (decode, update, log), shared by all detection models.

### Shared PTv3 backbone config
- Backbone definition extracted to `configs/tasks/segmentation3d/ptv3/backbone_base.yaml` (plus `backbone_tiny.yaml` overrides) and included by every PTv3 task `base`/`tiny` config (seg and det), removing the previously duplicated 25-line backbone block across configs.
- New segmentation `tiny.yaml` and the 128m/122m detection-range seg configs (`voxel005_128m`, `voxel012_122m`).

### PTv3 backbone cleanup
- `models/segmentation3d/backbones/ptv3.py`: removed the unused PDNorm (`pdnorm_*`) and classification-mode (`cls_mode`) parameters and branches (dead in this framework, which is single-dataset and dense-task only). `prepare_export_copy` renamed to `prepare_for_export` for a uniform export interface.

### Detection data pipeline
- **Datamodule** (`datamodule/t4dataset/detection3d.py`, `datamodule/nuscenes/detection3d.py`, shared helpers in `datamodule/common/detection3d.py`): T4 and NuScenes 3D detection datamodules built on the `collation_map` mechanism (per-key concat/stack/list strategy).
- **Repeat-factor sampling** (`datamodule/samplers.py`): `DistributedWeightedRandomSampler` plus frame-weight computation that oversamples rare-object frames, with a dedicated low-pedestrian bucket for small or sparse pedestrians.
- **3D box transforms** (`transforms/boxes3d/`): `LoadAnnotations3D` (box/label/num-points loading with name mapping) and box filters (`ObjectNameFilter`, `ObjectRangeFilter`, `ObjectMinPointsFilter`) that re-index all per-box arrays through a shared key set.
- **Point-cloud transforms**: `transforms/point_cloud/formatting.py` and `transforms/point_cloud/sweeps.py` (`LoadPointsFromMultiSweeps`), plus updates to crop, geometry, sampling, and segmentation formatting for the shared point pipeline.

### ONNX / TensorRT export rework
- **Per-module export** (`utils/deploy.py`): models expose `build_export_specs` returning a named per-module spec mapping (backbone and head or heads), deploy resolves it via `resolve_export_specs`, and the single-spec and dynamic-axes helpers were consolidated.
- **Graph modifiers** (`utils/onnx_modifiers.py`): `TransHeadTensorRTModifier` (with `TopKConstantKModifier` and `AttentionScaleToDivModifier`) makes the TransFusion head graph TensorRT-safe (static top-k, division-form attention scale).
- Per-head export prep via a uniform `prepare_for_export()` on each head, replacing a head-specific helper that was baked into the model.

---

## Framework-wide changes / minor updates

- **CLI checkpoint interface** (`cli/cli.py`, `cli/runtime.py`, `utils/cli/helpers.py`): replaced the old `+checkpoint=` Hydra override with explicit `--weights` (repeatable, for pretrained/partial init and multi-checkpoint merge) and `--resume-checkpoint` (full-state resume). Shell completion updated.
- **Checkpoint loading** (`utils/checkpoints.py`): the matching-weights loader raises on a name-matched but shape-mismatched tensor instead of silently skipping it. Unused legacy loaders removed.
- **Deploy** (`scripts/deploy.py`): requested-but-unsupported export stages now raise, and failures after MLflow-run creation mark the run FAILED.
- **Training/Optuna** (`scripts/train.py`): returns the configured `optimized_metric` for Hydra-Optuna sweeps.
- **MLflow** (`utils/mlflow_helpers.py`): warns (instead of silently recording "unknown") when git-SHA resolution fails, and applies stricter artifact-URI handling.
- **Default configs**: `pin_memory`/`persistent_workers` default to `true`, `strategy`/`devices` default to `auto`, and trainer/datamodule module defaults were updated.
- **spconv** (`ops/spconv/sparse_conv.py`): runtime `assert`s replaced with explicit exceptions so validation survives `python -O`.
- **Packaging**: `setup.py` build-spec `assert` replaced with a `RuntimeError` (guard fixed to run before use). `pyproject.toml` moves `pytest` into a `test` extra, and the pip `dev` extra pulls it in to mirror the pixi dev env.
- Calibration-status datamodule, tools, and util touch-ups for the shared interfaces.

---

## Tests

- New: detection datamodule, metric (`CenterDistanceMeanAP` including filtering, range, imperfect-curve, and DDP gather), task modules, losses, box transforms, ONNX modifiers, PTv3 detection model and export, and the deploy script.
- Updated: CLI, checkpoints, deploy utils, FRNet, PTv3 backbone, point-cloud and segmentation transforms, and the calibration datamodule.

## Docs

- `docs/models/ptv3.md` documents the detection task (head selection, configs, and train/test/deploy examples). Updates across CLI, configuration, deployment, training, mlflow, design, identity, adding-models, quickstart, devcontainer, and the FRNet and calibration model pages.

---

## Breaking changes / migration notes

- **Backbone/head deployment split (PTv3):** this PR splits the PTv3 export into a shared backbone graph plus separate head graph or graphs. Models now expose `build_export_specs` (plural) returning per-module specs (a standalone `backbone` and a separate `seg3d_head`/`det3d_head`), resolved by `resolve_export_specs`, replacing the previous single fused end-to-end ONNX (`build_export_spec`). This split is what lets one backbone export feed interchangeable heads. Runtime integrations that consumed the old single fused PTv3 graph must switch to the split backbone and head engines, and existing PTv3 exports/engines must be regenerated.
- **CLI:** `+checkpoint=...` is removed. Use `--weights` or `--resume-checkpoint`.
- **Checkpoint loading** is now strict on shape mismatch. A name-matched tensor with a different shape raises instead of being skipped.

## Checklist

- [ ] You've checked our [contribution overview](https://tier4.github.io/autoware-ml/contributing/contribution-overview/).
- [ ] Your PR follows our [pull request guidelines](https://tier4.github.io/autoware-ml/contributing/pull-request-guidelines/).

### CODE REVIEW

File: autoware_ml/models/segmentation3d/backbones/ptv3.py [hunk 1]
```diff
@@ -327,28 +328,23 @@ def _get_padding_and_inverse(
...
-                ] = pad[
-                    padded_bincount[0]
-                    - 2 * self.patch_size
-                    + (bincount[0] % self.patch_size) : padded_bincount[0] - self.patch_size
-                ]
+            n = shape_as_tensor(point.feat)[0].to(device=point.offset.device)
+            padded_n = ((n + self.patch_size - 1) // self.patch_size) * self.patch_size
+            unpad = torch.arange(n, device=point.offset.device)
+            # Clamp pad indices to [0, n-1]: dummy slots repeat the last real token.
+            pad = torch.clamp(torch.arange(padded_n, device=point.offset.device), max=n - 1)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/models/segmentation3d/backbones/ptv3.py:335]: irrelevant

File: autoware_ml/utils/point_cloud/structures.py [hunk 2]
```diff
@@ -123,3 +127,56 @@ def sparsify(self, pad: int = 96) -> None:
...
+        Inverse permutation of the same shape such that
+        ``result[i, order[i, j]] == j`` for all valid ``i`` and ``j``.
+    """
+    n = shape_as_tensor(order)[1]
+    src = torch.arange(n, device=order.device, dtype=order.dtype)
+    return torch.zeros_like(order).scatter_(
+        dim=1,
+        index=order,
+        src=src.unsqueeze(0).expand_as(order),
+    )
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/utils/point_cloud/structures.py:149]: irrelevant

File: autoware_ml/utils/deploy.py [hunk 3]
```diff
@@ -153,19 +151,51 @@ def infer_export_spec(model: L.LightningModule, batch: dict[str, Any]) -> Export
...
 
-    Models with deployment-specific wrappers override ``build_export_spec``
-    on :class:`~autoware_ml.models.base.BaseModel`. The default implementation
-    falls back to the generic forward-signature-based export path.
+    Returns:
+        Ordered mapping of module name to export specification.
     """
     batch = get_predict_batch(datamodule, model, device)
-    return model.build_export_spec(batch)
+    return model.build_export_specs(batch)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/utils/deploy.py:170]: https://github.com/tier4/autoware-ml/pull/43/commits/db86f5e989fa2f056d9834d4f1e99a354f582eb9

File: autoware_ml/configs/tasks/detection3d/ptv3/base.yaml [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+  grid_size: ${grid_size}
+  point_cloud_range: ${point_cloud_range}
+  metric:
+    _target_: autoware_ml.metrics.detection3d.CenterDistanceMeanAP
+    class_names: ${class_names}
+    thresholds: [0.5, 1.0, 2.0, 4.0]
+    filter_attributes:
+      - [bicycle, vehicle_state.parked]
+      - [bicycle, cycle_state.without_rider]
+      - [bicycle, motorcycle_state.without_rider]
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/configs/tasks/detection3d/ptv3/base.yaml]: https://github.com/tier4/autoware-ml/pull/43/commits/e41d0e5be4f2fdbcf92b58f7fb28e5eef867e647
